Posts/2021/03/File-System-Watcher-Engine-Event.md

```diff
@@ -0,0 +1,110 @@
+# A Reusable File System Event Watcher for PowerShell
+
+Have you ever needed a FilesystemWatcher and found it a bit cumbersome to interact with the
+[.Net frameworks FileSystemWatcher class](/dotnet/api/system.io.filesystemwatcher)
+within a powershell script?
+
+The PowerShell module
+[FSWatcherEngineEvent](https://www.powershellgallery.com/packages/FSWatcherEngineEvent) might help
+you out. The module provides commands for creating and managing file system watchers and will
+publish its notifications as PowerShell engine events instead of directly registered object events.
+
+After installing and importing of the module a new filesystem watcher can easily be created with the
+'New-FileSystemWatcher' command. As an example it will watch for changes in directory
+'C:\temp\files'. The command allows to specify the same parameters (with the same names) as if you
+are using a FileSystemWatcher directly. This includes the notification filters, file name filter and
+option to include subdirectories. Please refer to the Microsofts reference documentation of the
+FileSystemWatcher class for the details.
+
+```powershell
+PS> New-FileSystemWatcher -SourceIdentifier "myevent" -Path c:\temp\files
+```
+
+The filesystem watcher will now send notifications to PowerShells engine event queue using the
+source identifier "myevent". The event can be consumed by registering an event handler for the same
+source identifier. The following example just writes the whole event converted to JSON to the
+console:
+
+```powershell
+PS> Register-EngineEvent -SourceIdentifier "myevent" -Action { $event | ConvertTo-Json | Write-Host }
+
+Id     Name            PSJobTypeName   State         HasMoreData     Location             Command
+--     ----            -------------   -----         -----------     --------             -------
+1      myevent                         NotStarted    False                                $event|ConvertTo-Json|Wr…
+```
+
+To produce a new event some characters will be written to a file in the watched directory:
```

Comment:
To produce a new event, just write some characters to a file in the watched directory:

Reply:
fixed